NPE from DataCommunicator with lazy data binding

### Description of the bug

When one makes a lazy data binding (without providing count of items) to Grid, and scrolls quickly, there is an NPA thrown from JDK internals via a call trace coming from a DataCommunicator class. 

Interesting observations:

 * the NPE does NOT happen, if you scroll down slowly (Grid fetches more rows in smaller chunks 🤷‍♂️).
 * The NPE does NOT happen if the fetch callback collects the stream to a list and then again back to stream. See [the workaround](https://github.com/mstahv/bookstore-demo/commit/32aff57ffb8461276b3dc57d241a1938edb63ed8).

I haven't been able to figure out the true cause of the issue, but my guesses are: a bug in Vaadin Grid/DataCommunicator (I have a vague feeling I have seen a similar issue in some other Vaadin UI), bug in MicroStream, but in JDK.

### Expected behavior

Application should work :-)

### Minimal reproducible example

 * Check out the updated MicroStream example at [this specific commit](https://github.com/mstahv/bookstore-demo/tree/755a090accec9bbd733b0e39d6b7bc8b2a34ce19)
 * Run the example
 * Navigate to http://localhost:8080/inventory
 * Scroll the grid quickly downwards 

### Versions

Vaadin: 23.3.5
Flow: 23.3.3
Java: Eclipse Adoptium 19.0.1
OS: aarch64 Mac OS X 13.1
Browser: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36

Stacktrace:

```
java.lang.NullPointerException: null
	at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:67) ~[na:na]
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:500) ~[na:na]
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:484) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:542) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:567) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:670) ~[na:na]
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateParallel(ForEachOps.java:159) ~[na:na]
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateParallel(ForEachOps.java:173) ~[na:na]
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233) ~[na:na]
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596) ~[na:na]
	at com.vaadin.flow.data.provider.DataCommunicator.fetchFromProvider(DataCommunicator.java:1000) ~[flow-data-23.3.3.jar:23.3.3]
	at com.vaadin.flow.data.provider.DataCommunicator.activate(DataCommunicator.java:1440) ~[flow-data-23.3.3.jar:23.3.3]
	at com.vaadin.flow.data.provider.DataCommunicator.collectKeysToFlush(DataCommunicator.java:1405) ~[flow-data-23.3.3.jar:23.3.3]
	at com.vaadin.flow.data.provider.DataCommunicator.flush(DataCommunicator.java:1185) ~[flow-data-23.3.3.jar:23.3.3]
	at com.vaadin.flow.data.provider.DataCommunicator.lambda$requestFlush$7258256f$1(DataCommunicator.java:1112) ~[flow-data-23.3.3.jar:23.3.3]
	at com.vaadin.flow.internal.StateTree.lambda$runExecutionsBeforeClientResponse$2(StateTree.java:392) ~[flow-server-23.3.3.jar:23.3.3]
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183) ~[na:na]
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179) ~[na:na]
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625) ~[na:na]
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509) ~[na:na]
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499) ~[na:na]
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150) ~[na:na]
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173) ~[na:na]
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[na:na]
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596) ~[na:na]
	at com.vaadin.flow.internal.StateTree.runExecutionsBeforeClientResponse(StateTree.java:389) ~[flow-server-23.3.3.jar:23.3.3]
	at com.vaadin.flow.server.communication.UidlWriter.encodeChanges(UidlWriter.java:391) ~[flow-server-23.3.3.jar:23.3.3]
	at com.vaadin.flow.server.communication.UidlWriter.createUidl(UidlWriter.java:175) ~[flow-server-23.3.3.jar:23.3.3]
	at com.vaadin.flow.server.communication.UidlRequestHandler.createUidl(UidlRequestHandler.java:158) ~[flow-server-23.3.3.jar:23.3.3]
	at com.vaadin.flow.server.communication.UidlRequestHandler.writeUidl(UidlRequestHandler.java:146) ~[flow-server-23.3.3.jar:23.3.3]
	at com.vaadin.flow.server.communication.UidlRequestHandler.synchronizedHandleRequest(UidlRequestHandler.java:116) ~[flow-server-23.3.3.jar:23.3.3]
	at com.vaadin.flow.server.SynchronizedRequestHandler.handleRequest(SynchronizedRequestHandler.java:40) ~[flow-server-23.3.3.jar:23.3.3]
	at com.vaadin.flow.server.VaadinService.handleRequest(VaadinService.java:1564) ~[flow-server-23.3.3.jar:23.3.3]
	at com.vaadin.flow.server.VaadinServlet.service(VaadinServlet.java:369) ~[flow-server-23.3.3.jar:23.3.3]
	at com.vaadin.flow.spring.SpringServlet.service(SpringServlet.java:106) ~[vaadin-spring-23.3.3.jar:na]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:779) ~[tomcat-embed-core-9.0.71.jar:4.0.FR]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:227) ~[tomcat-embed-core-9.0.71.jar:9.0.71]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) ~[tomcat-embed-core-9.0.71.jar:9.0.71]
	at org.apache.catalina.core.ApplicationDispatcher.invoke(ApplicationDispatcher.java:711) ~[tomcat-embed-core-9.0.71.jar:9.0.71]
	at org.apache.catalina.core.ApplicationDispatcher.processRequest(ApplicationDispatcher.java:459) ~[tomcat-embed-core-9.0.71.jar:9.0.71]
	at org.apache.catalina.core.ApplicationDispatcher.doForward(ApplicationDispatcher.java:353) ~[tomcat-embed-core-9.0.71.jar:9.0.71]
	at org.apache.catalina.core.ApplicationDispatcher.forward(ApplicationDispatcher.java:313) ~[tomcat-embed-core-9.0.71.jar:9.0.71]
	at org.springframework.web.servlet.mvc.ServletForwardingController.handleRequestInternal(ServletForwardingController.java:141) ~[spring-webmvc-5.3.25.jar:5.3.25]
	at org.springframework.web.servlet.mvc.AbstractController.handleRequest(AbstractController.java:177) ~[spring-webmvc-5.3.25.jar:5.3.25]
	at org.springframework.web.servlet.mvc.SimpleControllerHandlerAdapter.handle(SimpleControllerHandlerAdapter.java:51) ~[spring-webmvc-5.3.25.jar:5.3.25]
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1071) ~[spring-webmvc-5.3.25.jar:5.3.25]
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:964) ~[spring-webmvc-5.3.25.jar:5.3.25]
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006) ~[spring-webmvc-5.3.25.jar:5.3.25]
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:909) ~[spring-webmvc-5.3.25.jar:5.3.25]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:696) ~[tomcat-embed-core-9.0.71.jar:4.0.FR]
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883) ~[spring-webmvc-5.3.25.jar:5.3.25]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:779) ~[tomcat-embed-core-9.0.71.jar:4.0.FR]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:227) ~[tomcat-embed-core-9.0.71.jar:9.0.71]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) ~[tomcat-embed-core-9.0.71.jar:9.0.71]
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53) ~[tomcat-embed-websocket-9.0.71.jar:9.0.71]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) ~[tomcat-embed-core-9.0.71.jar:9.0.71]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) ~[tomcat-embed-core-9.0.71.jar:9.0.71]
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100) ~[spring-web-5.3.25.jar:5.3.25]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) ~[spring-web-5.3.25.jar:5.3.25]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) ~[tomcat-embed-core-9.0.71.jar:9.0.71]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) ~[tomcat-embed-core-9.0.71.jar:9.0.71]
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93) ~[spring-web-5.3.25.jar:5.3.25]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) ~[spring-web-5.3.25.jar:5.3.25]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) ~[tomcat-embed-core-9.0.71.jar:9.0.71]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) ~[tomcat-embed-core-9.0.71.jar:9.0.71]
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201) ~[spring-web-5.3.25.jar:5.3.25]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) ~[spring-web-5.3.25.jar:5.3.25]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) ~[tomcat-embed-core-9.0.71.jar:9.0.71]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) ~[tomcat-embed-core-9.0.71.jar:9.0.71]
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:177) ~[tomcat-embed-core-9.0.71.jar:9.0.71]
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97) ~[tomcat-embed-core-9.0.71.jar:9.0.71]
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:541) ~[tomcat-embed-core-9.0.71.jar:9.0.71]
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:135) ~[tomcat-embed-core-9.0.71.jar:9.0.71]
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92) ~[tomcat-embed-core-9.0.71.jar:9.0.71]
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78) ~[tomcat-embed-core-9.0.71.jar:9.0.71]
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:360) ~[tomcat-embed-core-9.0.71.jar:9.0.71]
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:399) ~[tomcat-embed-core-9.0.71.jar:9.0.71]
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65) ~[tomcat-embed-core-9.0.71.jar:9.0.71]
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:891) ~[tomcat-embed-core-9.0.71.jar:9.0.71]
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1784) ~[tomcat-embed-core-9.0.71.jar:9.0.71]
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49) ~[tomcat-embed-core-9.0.71.jar:9.0.71]
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191) ~[tomcat-embed-core-9.0.71.jar:9.0.71]
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659) ~[tomcat-embed-core-9.0.71.jar:9.0.71]
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) ~[tomcat-embed-core-9.0.71.jar:9.0.71]
	at java.base/java.lang.Thread.run(Thread.java:1589) ~[na:na]
Caused by: java.lang.NullPointerException: Cannot store to object array because "this.curChunk" is null
	at java.base/java.util.stream.SpinedBuffer.accept(SpinedBuffer.java:263) ~[na:na]
	at java.base/java.util.stream.Streams$StreamBuilderImpl.accept(Streams.java:362) ~[na:na]
	at java.base/java.util.stream.Streams$StreamBuilderImpl.add(Streams.java:370) ~[na:na]
	at com.vaadin.flow.data.provider.DataCommunicator.lambda$fetchFromProvider$0(DataCommunicator.java:991) ~[flow-data-23.3.3.jar:23.3.3]
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183) ~[na:na]
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:1006) ~[na:na]
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509) ~[na:na]
	at java.base/java.util.stream.ForEachOps$ForEachTask.compute(ForEachOps.java:290) ~[na:na]
	at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:754) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1311) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1840) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1806) ~[na:na]
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177) ~[na:na]

```

--- a/flow-data/src/main/java/com/vaadin/flow/data/provider/DataCommunicator.java
+++ b/flow-data/src/main/java/com/vaadin/flow/data/provider/DataCommunicator.java
@@ -33,6 +33,9 @@ import java.util.function.Consumer;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 import com.vaadin.flow.component.Component;
 import com.vaadin.flow.component.ComponentUtil;
 import com.vaadin.flow.component.UI;
@@ -51,9 +54,6 @@ import com.vaadin.flow.internal.StateNode;
 import com.vaadin.flow.internal.StateTree;
 import com.vaadin.flow.shared.Registration;
 import com.vaadin.flow.shared.communication.PushMode;
-
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
 
 import elemental.json.Json;
 import elemental.json.JsonArray;
@@ -996,8 +996,19 @@ public class DataCommunicator<T> implements Serializable {
                 int page = 0;
                 do {
                     final int newOffset = offset + page * pageSize;
-                    doFetchFromDataProvider(newOffset, pageSize)
-                            .forEach(addItemAndCheckConsumer);
+                    Stream<T> dataProviderStream = doFetchFromDataProvider(
+                            newOffset, pageSize);
+                    // Stream.Builder is not thread safe, so for parallel stream
+                    // we need to first collect items before adding them
+                    if (dataProviderStream.isParallel()) {
+                        getLogger().debug(
+                                "Data provider {} has returned parallel stream on 'fetch' call",
+                                getDataProvider().getClass());
+                        dataProviderStream.collect(Collectors.toList())
+                                .forEach(addItemAndCheckConsumer);
+                    } else {
+                        dataProviderStream.forEach(addItemAndCheckConsumer);
+                    }
                     page++;
                 } while (page < pages
                         && fetchedPerPage.getAndSet(0) == pageSize);

--- a/flow-data/src/test/java/com/vaadin/flow/data/provider/DataCommunicatorAsyncTest.java
+++ b/flow-data/src/test/java/com/vaadin/flow/data/provider/DataCommunicatorAsyncTest.java
@@ -1,5 +1,7 @@
 package com.vaadin.flow.data.provider;
 
+import java.util.Arrays;
+import java.util.Collection;
 import java.util.List;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.Executor;
@@ -12,6 +14,8 @@ import java.util.stream.Stream;
 import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
 import org.mockito.Mock;
 import org.mockito.MockitoAnnotations;
 
@@ -27,6 +31,7 @@ import com.vaadin.flow.shared.communication.PushMode;
 
 import elemental.json.JsonValue;
 
+@RunWith(Parameterized.class)
 public class DataCommunicatorAsyncTest {
 
     /**
@@ -84,6 +89,18 @@ public class DataCommunicatorAsyncTest {
     public Range lastClear = null;
     public Range lastSet = null;
     public int lastUpdateId = -1;
+
+    private final boolean dataProviderWithParallelStream;
+
+    public DataCommunicatorAsyncTest(boolean dataProviderWithParallelStream) {
+        this.dataProviderWithParallelStream = dataProviderWithParallelStream;
+    }
+
+    @Parameterized.Parameters
+    public static Collection<Boolean> testParameters() {
+        // Runs tests with both sequential and parallel data provider streams
+        return Arrays.asList(false, true);
+    }
 
     @Before
     public void init() {
@@ -197,12 +214,18 @@ public class DataCommunicatorAsyncTest {
                 // setting lastSet here
                 lastSet = Range.withLength(query.getOffset(), query.getLimit());
                 latch.countDown();
-                return IntStream
-                        .range(query.getOffset(),
-                                query.getLimit() + query.getOffset())
+                return asParallelIfRequired(IntStream.range(query.getOffset(),
+                        query.getLimit() + query.getOffset()))
                         .mapToObj(Item::new);
             }
         };
+    }
+
+    private IntStream asParallelIfRequired(IntStream stream) {
+        if (dataProviderWithParallelStream) {
+            return stream.parallel();
+        }
+        return stream;
     }
 
     private void fakeClientCommunication() {

--- a/flow-data/src/test/java/com/vaadin/flow/data/provider/DataCommunicatorTest.java
+++ b/flow-data/src/test/java/com/vaadin/flow/data/provider/DataCommunicatorTest.java
@@ -16,6 +16,8 @@
 package com.vaadin.flow.data.provider;
 
 import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
 import java.util.List;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
@@ -25,6 +27,19 @@ import java.util.concurrent.locks.ReentrantLock;
 import java.util.stream.Collectors;
 import java.util.stream.IntStream;
 import java.util.stream.Stream;
+
+import org.hamcrest.CoreMatchers;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
 
 import com.vaadin.flow.component.Component;
 import com.vaadin.flow.component.ComponentEventListener;
@@ -39,19 +54,10 @@ import com.vaadin.flow.internal.StateNode;
 import com.vaadin.flow.server.VaadinRequest;
 import com.vaadin.flow.server.VaadinService;
 import com.vaadin.flow.server.VaadinSession;
-import org.hamcrest.CoreMatchers;
-import org.junit.Assert;
-import org.junit.Before;
-import org.junit.Rule;
-import org.junit.Test;
-import org.junit.rules.ExpectedException;
-import org.mockito.ArgumentCaptor;
-import org.mockito.Mock;
-import org.mockito.Mockito;
-import org.mockito.MockitoAnnotations;
 
 import elemental.json.JsonValue;
 
+@RunWith(Parameterized.class)
 public class DataCommunicatorTest {
     /**
      * Test item that uses id for identity.
@@ -109,6 +115,18 @@ public class DataCommunicatorTest {
     public Range lastSet = null;
     public int lastUpdateId = -1;
     private int pageSize;
+
+    private final boolean dataProviderWithParallelStream;
+
+    public DataCommunicatorTest(boolean dataProviderWithParallelStream) {
+        this.dataProviderWithParallelStream = dataProviderWithParallelStream;
+    }
+
+    @Parameterized.Parameters
+    public static Collection<Boolean> testParameters() {
+        // Runs tests with both sequential and parallel data provider streams
+        return Arrays.asList(false, true);
+    }
 
     @Before
     public void init() {
@@ -1655,7 +1673,8 @@ public class DataCommunicatorTest {
                     count -= delta;
                     modifiedCount = true;
                 }
-                return IntStream.range(query.getOffset(), count)
+                return asParallelIfRequired(
+                        IntStream.range(query.getOffset(), count))
                         .mapToObj(Item::new);
             }
         };
@@ -1675,9 +1694,8 @@ public class DataCommunicatorTest {
 
             @Override
             public Stream<Item> fetch(Query<Item, Object> query) {
-                return IntStream
-                        .range(query.getOffset(),
-                                query.getLimit() + query.getOffset())
+                return asParallelIfRequired(IntStream.range(query.getOffset(),
+                        query.getLimit() + query.getOffset()))
                         .mapToObj(Item::new);
             }
         };
@@ -1698,10 +1716,18 @@ public class DataCommunicatorTest {
             @Override
             public Stream<Item> fetch(Query<Item, Object> query) {
                 int end = Math.min(query.getRequestedRangeEnd(), size);
-                return IntStream.range(query.getOffset(), end)
+                return asParallelIfRequired(
+                        IntStream.range(query.getOffset(), end))
                         .mapToObj(Item::new);
             }
         };
+    }
+
+    private IntStream asParallelIfRequired(IntStream stream) {
+        if (dataProviderWithParallelStream) {
+            return stream.parallel();
+        }
+        return stream;
     }
 
     public static class MockUI extends UI {

